### PR TITLE
added support for selecting the counted card

### PR DESCRIPTION
**Pull request by joelong01**
there are a lot of comments that talk about strategy and things to improve.

let me know if the use of rust looks right.

### CODE REVIEW

File: game/src/select_cards.rs [hunk 1]
```diff
@@ -49,13 +52,343 @@ fn get_crib_cards(hand: &[Card], held_cards: &[Card]) -> Vec<Card> {
...
+ *
+ *  my strategy for this is to implement it the way I play the game.  by its nature it is a bunch of "if this then that" calls, so this function
+ *  tends to be very long and very experimental.  I have added logic in the past and then tested it by playing millions (literally) of games of
+ *  one algorithm against the other to see if it really works (it takes about 20 seconds to run a million games on a beefy PC...)
+ *
+ *
+ */
+pub fn get_next_counted_card(
+    counted_cards: Vec<Card>,
+    mut cards_left: Vec<Card>, // needs to be mut because we .sort() it
```
Comment (DanMass) [inline game/src/select_cards.rs:73]: would it be better to `clone` and then `sort` inside the fn or is the mut ok?

Comment (DanMass) [reply game/src/select_cards.rs:73]: For future proofing, shouldn't this fn take in more state, like other cards we've already seen (what I cribbed and the starter) and how many cards my opponent has left?

Comment (DanMass) [reply game/src/select_cards.rs:73]: I probably also want the record of who played each of the counted_cards?

File: game/src/select_cards.rs [hunk 2]
```diff
@@ -49,13 +52,343 @@ fn get_crib_cards(hand: &[Card], held_cards: &[Card]) -> Vec<Card> {
...
+    //
+    //  we only want to interate through the 2 card combinations once, so
+    //  we set a strategic weight to pick between the various scenarios and
+    //  set the card_to_play when we find one
+    //
+    let mut strategic_weight = 0;
+    let mut card_to_play: Card = Card::new(Rank::Unknown, Suit::Unknown);
+
+    for mut cards in two_card_combi {
+        if cards.len() != 2 {
```
Comment (DanMass) [inline game/src/select_cards.rs:146]: This is a weird panic. At some point, we should trust the code above, that we own, to do what it says it does.

File: game/src/select_cards.rs [hunk 3]
```diff
@@ -49,13 +52,343 @@ fn get_crib_cards(hand: &[Card], held_cards: &[Card]) -> Vec<Card> {
...
+    mut cards_left: Vec<Card>, // needs to be mut because we .sort() it
+) -> Result<Card, Box<dyn Error>> {
+    let current_count: i32 = counted_cards.iter().map(|c| c.value).sum::<i32>();
+    //
+    //  if you only have one card left, play it if you can
+    if cards_left.len() == 1 {
+        if current_count + cards_left[0].value <= 31 {
+            return Ok(cards_left[0]);
+        } else {
+            return Err("Go".into());
```
Comment (DanMass) [inline game/src/select_cards.rs:82]: It feels like Go should be a valid result, not a kind of Error.

File: game/src/select_cards.rs [hunk 4]
```diff
@@ -49,13 +52,343 @@ fn get_crib_cards(hand: &[Card], held_cards: &[Card]) -> Vec<Card> {
...
+ *  my strategy for this is to implement it the way I play the game.  by its nature it is a bunch of "if this then that" calls, so this function
+ *  tends to be very long and very experimental.  I have added logic in the past and then tested it by playing millions (literally) of games of
+ *  one algorithm against the other to see if it really works (it takes about 20 seconds to run a million games on a beefy PC...)
+ *
+ *
+ */
+pub fn get_next_counted_card(
+    counted_cards: Vec<Card>,
+    mut cards_left: Vec<Card>, // needs to be mut because we .sort() it
+) -> Result<Card, Box<dyn Error>> {
```
Comment (DanMass) [inline game/src/select_cards.rs:74]: Better semantics might be Option<Card> and the semantics of the call are either I have a card to play or I don't. Some(...) is a play and None is a Go. I wouldn't treat Go as an Error, given that it's a thing that we expect to happen.

File: game/src/select_cards.rs [hunk 5]
```diff
@@ -49,13 +52,343 @@ fn get_crib_cards(hand: &[Card], held_cards: &[Card]) -> Vec<Card> {
...
+    //
+    //  one of the things I want to explore is defeating the "get a 15" strategy above. the other common strategy (which we use below) is to start with a card
+    //  that you have a pair...the hope is the apponent pairs the card, so you can get 3 of a kind (e.g. play a 4, opponent plays a 4, you play your second 4)
+    //  would it be better to not pair the card and give up the two points?  over the course of the game, stopping 3 of a kind might be worth it.
+    //
+    //  this is an area ripe for innovation!
+    //
+
+    //
+    //  Note: I need the card to play for the max score - i tried to do this in a variety of ways using map/fold and I could get the max score, but the card
```
Comment (DanMass) [inline game/src/select_cards.rs:104]: You could accumulate to a tuple of `(i32, Card)` with map/fold, so you don't lose the card. I think that _would_ be cleaner.

File: game/src/select_cards.rs [hunk 6]
```diff
@@ -49,13 +52,343 @@ fn get_crib_cards(hand: &[Card], held_cards: &[Card]) -> Vec<Card> {
...
+    //  get the most points - innovate here and pick the *best* one, not necessarily the one with most points!
+    if max > 0 {
+        return Ok(card_to_play);
+    }
+    //
+    //  we can play, but we can't get points.
+
+    //
+    //  find all combinations of 2 cards in the cards I have left...
+    let two_card_combi = all_combinations_of_size(playable_cards, 2, 2);
```
Comment (DanMass) [inline game/src/select_cards.rs:136]: Do you need the var, if you're only using the call to iterate over below?

File: game/src/select_cards.rs [hunk 7]
```diff
@@ -49,13 +52,343 @@ fn get_crib_cards(hand: &[Card], held_cards: &[Card]) -> Vec<Card> {
...
+    //  set the card_to_play when we find one
+    //
+    let mut strategic_weight = 0;
+    let mut card_to_play: Card = Card::new(Rank::Unknown, Suit::Unknown);
+
+    for mut cards in two_card_combi {
+        if cards.len() != 2 {
+            panic!("all_combinations_of_min_size returned the wrong size Vec!");
+        }
+        cards.sort();   // this sort might not be needed, but i'm not sure if all_combinations_of_size guarantees to returns sorted if the input was sorted
```
Comment (DanMass) [inline game/src/select_cards.rs:149]: That code is correct that it  _is_. needed. From the Combinator docs...

/// The order of the elements _within_ each combination is stable and consistent
/// with their order in `vector`.

File: game/src/select_cards.rs [hunk 8]
```diff
@@ -49,13 +52,343 @@ fn get_crib_cards(hand: &[Card], held_cards: &[Card]) -> Vec<Card> {
...
+    //  Another innovation idea to test is that a human player will figure out what the game is always doing -- so, for example, it is easy
+    //  to figure out that if the computer starts with (say) a 8, then they are probably trying to match a pair.  if they start with a card
+    //  less than 5, they are trying to get to 15.  this extra information can be leveraged to help the player.  So instead we should figure
+    //  out what valid cards could be played and then if then set strategic_weight to be the same for adding up to 15 and for pairs and then
+    //  randomly pick which one to use.
+    //
+
+    if strategic_weight > 0 {
+        if card_to_play.rank == Rank::Unknown {
+            panic!("strategic weight is set but the card_to_play is not not");
```
Comment (DanMass) [inline game/src/select_cards.rs:223]: not not?

File: game/src/select_cards.rs [hunk 9]
```diff
@@ -49,13 +52,343 @@ fn get_crib_cards(hand: &[Card], held_cards: &[Card]) -> Vec<Card> {
...
+    //
+    if cards_left[cards_left.len() - 1].value != 5 {
+        return Ok(cards_left[cards_left.len() - 1]);
+    }
+    //
+    //  if it is a 5, return the next highest one -- we know it is not a pair because if it was, we'd have played it based on what we have above
+    //  we also know that we have at least two cards, because if there was only one, we would have already picked it.
+    Ok(cards_left[cards_left.len() - 2])
+}
+#[cfg(test)]
```
Comment (DanMass) [inline game/src/select_cards.rs:238]: leave a blank line

File: game/src/select_cards.rs [hunk 10]
```diff
@@ -49,13 +52,343 @@ fn get_crib_cards(hand: &[Card], held_cards: &[Card]) -> Vec<Card> {
...
+                            "unexpected card returned: {}.  expected: {}",
+                            card.name(),
+                            $expected_card.name()
+                        );
+                    }
+                }
+            }
+        };
+    }
+    /*
```
Comment (DanMass) [inline game/src/select_cards.rs:277]: blank line

we should get consistent on comment style. If we wanted to go with `///`, does Rust do the right thing with a Doc comment over a macro like this?

**Review by DanMass — APPROVED** [review]
The rules are definitely something to think about, both in terms of how to express them and the readability of that long method if the different strategies were broken out into their own fns, so that the higher level progression was clearer to the reader.

File: game/src/select_cards.rs (hunk 6 above)
Comment (joelong01) [reply game/src/select_cards.rs:136]: probably not, but it makes it easier if you ever want to use it...it also makes it so I can add a comment easier.

File: game/src/select_cards.rs (hunk 7 above)
Comment (joelong01) [reply game/src/select_cards.rs:149]: deleted comment

**Pull request merged**